plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,234 @@
+#include "irods/rcConnect.h"
+#include "irods/rodsPackInstruct.h"
+#include "irods/rcMisc.h"
+#include "irods/client_api_allowlist.hpp"
+
+#include "irods/apiHandler.hpp"
+
+#include <functional>
+
+#ifdef RODS_SERVER
+
+//
+// Server-side Implementation
+//
+
+#  include "irods/switch_user.h"
+#  include "irods/irods_rs_comm_query.hpp"
+#  include "irods/rodsErrorTable.h"
+#  include "irods/irods_logger.hpp"
+#  include "irods/server_utilities.hpp"
+
+#  define IRODS_USER_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+#  include "irods/user_administration.hpp"
+
+#  include <cstring>
+#  include <string>
+#  include <string_view>
+
+// This global variable is required so that the API plugin can access RcComm connections
+// created as a result of previous API calls (i.e. connections created due to redirection).
+extern zoneInfo* ZoneInfoHead;
+
+namespace
+{
+    //
+    // Namespace and Type aliases
+    //
+
+    using log_api = irods::experimental::log::api;
+
+    //
+    // Function Prototypes
+    //
+
+    auto call_switch_user(irods::api_entry*, RsComm*, SwitchUserInput*) -> int;
+    auto rs_switch_user(RsComm*, SwitchUserInput*) -> int;
+    auto check_input(const SwitchUserInput*) -> int;
+    auto update_server_to_server_connections(RsComm&) -> void;
+
+    //
+    // Function Implementations
+    //
+
+    auto call_switch_user(irods::api_entry* _api, RsComm* _comm, SwitchUserInput* _input) -> int
+    {
+        return _api->call_handler<SwitchUserInput*>(_comm, _input);
+    } // call_switch_user
+
+    auto check_input(const SwitchUserInput* _input) -> int
+    {
+        if (!_input) {
+            log_api::error("Invalid input argument: received a null pointer");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        if (std::strlen(_input->username) == 0) {
+            log_api::error("Invalid username argument: null or empty");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        if (std::strlen(_input->zone) == 0) {
+            log_api::error("Invalid zone argument: null or empty");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        return 0;
+    } // check_input
+
+    auto update_server_to_server_connections(RsComm& _comm) -> void
+    {
+        const auto update_zone_connection = [&client = _comm.clientUser](rodsServerHost& _host)
+        {
+            if (_host.conn) {
+                const auto ec = rc_switch_user(_host.conn, client.userName, client.rodsZone);
+                log_api::info("rc_switch_user ec = [{}]", ec);
+            }
+        };
+
+        for (auto* zone_ptr = ZoneInfoHead; zone_ptr; zone_ptr = zone_ptr->next) {
+            for (auto* host_ptr = zone_ptr->primaryServerHost; host_ptr; host_ptr = host_ptr->next) {
+                update_zone_connection(*host_ptr);
+            }
+
+            for (auto* host_ptr = zone_ptr->secondaryServerHost; host_ptr; host_ptr = host_ptr->next) {
+                update_zone_connection(*host_ptr);
+            }
```

Comment:
Consider what happens to the entries in `ZoneInfoHead` upon invoking `rcDisconnect` on them.

Perhaps, we don't have to do anything.

Reply:
Based on what I've read, it looks like we don't have to do anything more in regard to `ZoneInfoHead`. `cleanupAndExit()` calls `disconnectAllSvrToSvrConn()` which iterates over the linked list pointed to by `ServerHostHead` and invokes `rcDisconnect()` on each entry.

This may also mean that the code snippet referenced here can iterate over the `ServerHostHead` instead of the `ZoneInfoHead`.

Reply:
I wrote a little program that does the following:
1. Connects to consumer 1 as alice.
2. Creates a data object on otherResc under alice's home collection.
3. Switches user identity to bobby via `rc_switch_user`.
4. Creates a data object on the default resource under bobby's home collection.

Next, I used the testing environment to launch a zone with the following config:
- 1 provider
- 2 consumers
- 3 users (rods, alice, and bobby)
- 2 resources (demoResc and otherResc)

With this setup, I then ran the test program from consumer 1.

This test verifies that `rc_switch_user` updates the server-to-server connections as expected.

It seems that things are working as expected.

Do we feel more testing is needed?

Having documentation around `ZoneInfoHead` and friends would be beneficial here. I'll create an issue for this.

Reply:
Oh snap! @alanking already documented the relationship between `ZoneInfoHead` and `ServerHostHead`. No issue needed.

https://github.com/irods/irods/blob/54c2d73cab1edaff93ae61bc008eeb4361ff4b98/server/core/include/irods/rodsConnect.h#L115-L128

Based on the comments, we're good to mark this discussion as resolved.

Reply:
This does infer that changing things to use `ServerHostHead` is acceptable.

This is very helpful. Found one typo in last statement.
https://github.com/irods/irods/blob/a86b0ed40bcb2bd23cf526a52244286b78285468/server/core/include/irods/rsGlobalExtern.hpp#L34-L47

Reply:
I'm going to leave the code as is.

`secondaryServerHost` is never populated in iRODS so it is essentially a NOP. However, if it suddenly begins to be populated, then the API plugin will update those connections too.

Reply:
makes sense to me!